installer/events.go

```diff
@@ -0,0 +1,273 @@
+				i.logger.Debug(fmt.Sprintf("GetEventsSince Prompt Scan Error: %s", err.Error()))
+				continue
+			}
+			event.Prompt = p
+		}
+		events = append(events, event)
+	}
+	return events
+}
+
+func (i *Installer) SendEvent(event *Event) {
+	event.Timestamp = time.Now()
+	event.ID = fmt.Sprintf("event-%d", event.Timestamp.UnixNano())
+
+	if event.Type == "prompt" {
+		if event.Prompt == nil {
+			i.logger.Debug(fmt.Sprintf("SendEvent Error: Invalid prompt event: %v", event))
+			return
+		}
+		event.PromptID = event.Prompt.ID
+	}
+
+	if event.Type == "error" {
+		i.logger.Error(fmt.Sprintf("Error: %s", event.Description))
+	} else {
+		i.logger.Info(fmt.Sprintf("Event: %s: %s", event.Type, event.Description))
+	}
+
+	err := i.dbInsertItem("events", event)
+	if err != nil {
+		i.logger.Debug(err.Error())
+	}
+
+	for _, sub := range i.subscriptions {
+		go sub.SendEvents(i)
+	}
+}
+
+func (c *BaseCluster) findPrompt(id string) (*Prompt, error) {
+	if c.pendingPrompt != nil && c.pendingPrompt.ID == id {
+		return c.pendingPrompt, nil
+	}
+	return nil, errors.New("Prompt not found")
+}
+
+func (c *BaseCluster) sendPrompt(prompt *Prompt) *Prompt {
+	c.pendingPrompt = prompt
+
+	if err := c.installer.dbInsertItem("prompts", prompt); err != nil {
+		c.installer.logger.Debug(fmt.Sprintf("sendPrompt db insert error: %s", err.Error()))
+	}
+
+	c.sendEvent(&Event{
+		Type:      "prompt",
+		ClusterID: c.ID,
+		Prompt:    prompt,
+	})
+
+	res := <-prompt.resChan
+	prompt.Resolved = true
+	prompt.Yes = res.Yes
+	prompt.Input = res.Input
+
+	if err := c.dbUpdatePrompt(prompt); err != nil {
+		c.installer.logger.Debug(fmt.Sprintf("sendPrompt db update error: %s", err.Error()))
+		return res
+	}
+
+	c.sendEvent(&Event{
+		Type:      "prompt",
+		ClusterID: c.ID,
+		Prompt:    prompt,
+	})
+
+	return res
+}
+
+func (c *BaseCluster) dbUpdatePrompt(prompt *Prompt) error {
+	c.installer.dbMtx.Lock()
+	defer c.installer.dbMtx.Unlock()
+
+	tx, err := c.installer.db.Begin()
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(`UPDATE prompts SET Resolved = $1, Yes = $2, Input = $3 WHERE ID == $4`, prompt.Resolved, prompt.Yes, prompt.Input, prompt.ID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (i *Installer) dbInsertItem(tableName string, item interface{}) error {
+	i.dbMtx.Lock()
+	defer i.dbMtx.Unlock()
+
+	fields, err := ql.Marshal(item)
+	if err != nil {
+		return err
+	}
+	vStr := make([]string, 0, len(fields))
+	for idx := range fields {
+		vStr = append(vStr, fmt.Sprintf("$%d", idx+1))
+	}
+	list, err := ql.Compile(fmt.Sprintf(`
+    INSERT INTO %s VALUES(%s);
+	`, tableName, strings.Join(vStr, ", ")))
+	if err != nil {
+		return err
+	}
+	tx, err := i.db.Begin()
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(list.String(), fields...)
```

Comment:
Looks like there are a few other similar transactions that only have one statement.